samza-core/src/main/scala/org/apache/samza/checkpoint/OffsetManager.scala

```diff
@@ -231,24 +231,64 @@ class OffsetManager(
-        Map[SystemStreamPartition, String]()
-      }
+      val taskStartingOffsets = startingOffsets.getOrElse(taskName,
+        throw new SamzaException("Couldn't find starting offsets for task: " + taskName))
+
+      val taskSSPs = systemStreamPartitions.getOrElse(taskName,
+        throw new SamzaException("No SSPs registered for task: " + taskName))
 
-      new Checkpoint(new HashMap(partitionOffsets.asJava)) // Copy into new Map to prevent mutation
+      // filter the offsets in case the task model changed since the last checkpoint was written.
+      val taskLastProcessedOffsets =
+        lastProcessedOffsets.getOrDefault(taskName, new ConcurrentHashMap()).asScala
+          .filterKeys(taskSSPs.contains)
+
+      val modifiedTaskOffsets = getModifiedOffsets(taskStartingOffsets, taskLastProcessedOffsets)
+      new Checkpoint(new HashMap(modifiedTaskOffsets)) // Copy into new Map to prevent mutation
     } else {
       debug("Returning null checkpoint for taskName %s because no checkpoint manager/callback is defined." format taskName)
       null
     }
   }
 
+  /**
+    * Call the beforeCheckpoint() method on [[CheckpointListener]] SystemConsumers to give them a chance to inspect and
+    * potentially modify the offsets about to be checkpointed.
+    */
+  def getModifiedOffsets(taskStartingOffsets: Map[SystemStreamPartition, String],
+      taskLastProcessedOffsets: Map[SystemStreamPartition, String]): HashMap[SystemStreamPartition, String] = {
+    val modifiedOffsets = new HashMap[SystemStreamPartition, String]()
+    modifiedOffsets.putAll(taskLastProcessedOffsets.asJava)
+
+    taskLastProcessedOffsets
+      .groupBy { case (ssp, offset) => ssp.getSystem }  // Map(systemName -> Map(ssp -> offset))
+      .foreach { case (systemName, systemSSPLastProcessedOffsets) =>
+      if (checkpointListeners.contains(systemName)) {
+        val checkpointListener = checkpointListeners(systemName)
+
+        // if we haven't processed any messages from the system since we started, we don't need to call
+        // the checkpoint listener. also, in case of Kafka, getting the safe offset is only possible after
+        // first poll() is successful. check if we need to adjust the offset for any of the ssp in the system.
+        val needModifiedOffsets = systemSSPLastProcessedOffsets.exists { case (ssp, sspLastProcessedOffset) =>
+          // starting offsets are checkpointed offsets + 1 (see 'loadStartingOffsets'),
+          // while on container start last processed offsets are equal to checkpointed offsets.
+          // subtract one from starting offsets to compare them with last processed offsets.
+          val adjustedSSPStartingOffset = taskStartingOffsets.get(ssp) match {
+            case Some(offset) if offset != null => (offset.toLong - 1).toString
+            case _ => ""
+          }
+          !adjustedSSPStartingOffset.equals(sspLastProcessedOffset)
+        }
+
+        if (needModifiedOffsets) {
```

Comment:
Please correct me if i'm wrong here.

This is more like a optimization right, why can't we bluntly call the listener.beforeCheckpoint without the needModifiedOffsets check. For most of the I/0 systems, checkpoint listener implementation will simply be a passthrough filter. For kafka system alone, this optimization might save few kafka API calls before we commit the first checkpoint.

Reply:
See discussion with Jagadish above.